<a href="https://colab.research.google.com/github/annakoskina/lab3_python/blob/master/lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Установить через pip необходимые пакеты**

In [0]:
!pip install psycopg2-binary

# **Подключение к БД PostgreSQL на основе sqlalchemy**

In [35]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sqlalchemy import create_engine

#Строка подключения - из Heroku
postgres_str='postgres://gwdpbzbuplmryj:de370901fe7793ba3aba4efc3cefbccc3d28fd167f77dbbcb16af31eaea3ca1c@ec2-54-247-177-254.eu-west-1.compute.amazonaws.com:5432/der83ujo7c7j00'

#Создание курсора для работы с БД
cnx = create_engine(postgres_str)

#Запрос на получение всех таблиц из БД
pd.read_sql_query('SELECT * FROM pg_catalog.pg_tables', cnx)

schemaname                tablename  ... hastriggers rowsecurity
0               public                 students  ...       False       False
1               public                entrexams  ...       False       False
2               public                playstore  ...       False       False
3           pg_catalog             pg_statistic  ...       False       False
4               public            semesterexams  ...       False       False
5           pg_catalog         pg_foreign_table  ...       False       False
6           pg_catalog                pg_authid  ...       False       False
7           pg_catalog          pg_user_mapping  ...       False       False
8           pg_catalog          pg_subscription  ...       False       False
9           pg_catalog           pg_largeobject  ...       False       False
10          pg_catalog                  pg_type  ...       False       False
11          pg_catalog             pg_attribute  ...       False       False
12          pg_catalog                  pg_proc  ...       False       False
13          pg_catalog                 pg_class  ...       False       False
14          pg_catalog               pg_attrdef  ...       False       False
15          pg_catalog            pg_constraint  ...       False       False
16          pg_catalog              pg_inherits  ...       False       False
17          pg_catalog                 pg_index  ...       False       False
18          pg_catalog              pg_operator  ...       False       False
19          pg_catalog              pg_opfamily  ...       False       False
20          pg_catalog               pg_opclass  ...       False       False
21          pg_catalog                    pg_am  ...       False       False
22          pg_catalog                  pg_amop  ...       False       False
23          pg_catalog                pg_amproc  ...       False       False
24          pg_catalog              pg_language  ...       False       False
25          pg_catalog  pg_largeobject_metadata  ...       False       False
26          pg_catalog             pg_aggregate  ...       False       False
27          pg_catalog         pg_statistic_ext  ...       False       False
28          pg_catalog               pg_rewrite  ...       False       False
29          pg_catalog               pg_trigger  ...       False       False
..                 ...                      ...  ...         ...         ...
43          pg_catalog         pg_shdescription  ...       False       False
44          pg_catalog             pg_ts_config  ...       False       False
45          pg_catalog         pg_ts_config_map  ...       False       False
46          pg_catalog               pg_ts_dict  ...       False       False
47          pg_catalog             pg_ts_parser  ...       False       False
48          pg_catalog           pg_ts_template  ...       False       False
49          pg_catalog             pg_extension  ...       False       False
50          pg_catalog  pg_foreign_data_wrapper  ...       False       False
51          pg_catalog        pg_foreign_server  ...       False       False
52          pg_catalog                pg_policy  ...       False       False
53          pg_catalog    pg_replication_origin  ...       False       False
54          pg_catalog           pg_default_acl  ...       False       False
55          pg_catalog            pg_init_privs  ...       False       False
56          pg_catalog              pg_seclabel  ...       False       False
57          pg_catalog            pg_shseclabel  ...       False       False
58          pg_catalog             pg_collation  ...       False       False
59          pg_catalog     pg_partitioned_table  ...       False       False
60          pg_catalog                 pg_range  ...       False       False
61          pg_catalog             pg_transform  ...       False       False
62          pg_catalog              pg_sequence  ...       False       False
63      

# **Загрузка данных из CSV -файла**

In [0]:
df = pd.read_csv("googleplaystore.csv")

In [39]:
df.head()# Заголовок и первые 10 записей

App  ...   Android Ver
0     Photo Editor & Candy Camera & Grid & ScrapBook  ...  4.0.3 and up
1                                Coloring book moana  ...  4.0.3 and up
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ...  4.0.3 and up
3                              Sketch - Draw & Paint  ...    4.2 and up
4              Pixel Draw - Number Art Coloring Book  ...    4.4 and up

[5 rows x 13 columns]

# **Загрузка датафрейма с даными в БД в таблицу playstore**

In [0]:
df.to_sql("playstore", con=cnx)

# **Запрос на выборку из БД**

In [40]:
q=pd.read_sql_query('SELECT * FROM playstore;', cnx)

q

index  ...         Android Ver
0         0  ...        4.0.3 and up
1         1  ...        4.0.3 and up
2         2  ...        4.0.3 and up
3         3  ...          4.2 and up
4         4  ...          4.4 and up
5         5  ...          2.3 and up
6         6  ...        4.0.3 and up
7         7  ...          4.2 and up
8         8  ...          3.0 and up
9         9  ...        4.0.3 and up
10       10  ...          4.1 and up
11       11  ...          4.0 and up
12       12  ...          4.1 and up
13       13  ...          4.4 and up
14       14  ...          2.3 and up
15       15  ...          4.2 and up
16       16  ...          4.1 and up
17       17  ...          2.3 and up
18       18  ...        4.0.3 and up
19       19  ...          4.1 and up
20       20  ...          4.1 and up
21       21  ...        4.0.3 and up
22       22  ...        4.0.3 and up
23       23  ...          4.1 and up
24       24  ...          4.1 and up
25       25  ...          3.0 and up
26       26  ...        4.0.3 and up
27       27  ...        4.0.3 and up
28       28  ...          2.3 and up
29       29  ...          2.3 and up
...     ...  ...                 ...
5787   5787  ...          4.0 and up
5788   5788  ...          4.1 and up
5789   5789  ...          4.1 and up
5790   5790  ...          3.0 and up
5791   5791  ...          3.0 and up
5792   5792  ...          2.2 and up
5793   5793  ...          4.1 and up
5794   5794  ...          4.2 and up
5795   5795  ...          4.1 and up
5796   5796  ...          4.1 and up
5797   5797  ...          4.1 and up
5798   5798  ...          4.4 and up
5799   5799  ...          4.4 and up
5800   5800  ...          4.3 and up
5801   5801  ...          3.1 and up
5802   5802  ...          4.1 and up
5803   5803  ...          4.1 and up
5804   5804  ...        4.0.3 and up
5805   5805  ...          2.1 and up
5806   5806  ...        4.0.3 and up
5807   5807  ...          4.0 and up
5808   5808  ...          4.3 and up
5809   5809  ...        4.0.3 and up
5810   5810  ...        4.0.3 and up
5811   5811  ...        2.3.3 and up
5812   5812  ...          2.3 and up
5813   5813  ...          4.1 and up
5814   5814  ...          2.3 and up
5815   5815  ...        4.0.3 and up
5816   5816  ...  Varies with device

[5817 rows x 14 columns]

# **Просмотр структуры полей БД средствами Pandas**

In [0]:
q.columns

Index([u'index', u'App', u'Category', u'Rating', u'Reviews', u'Size',
       u'Installs', u'Type', u'Price', u'Content Rating', u'Genres',
       u'Last Updated', u'Current Ver', u'Android Ver'],
      dtype='object')

# **Рассмотрим гипотезу, что количество просмотров зависит от количества инсталляций. Исследование проводить для данных, размер приложений которых меньше 20M. Для подтверждения этой гипотезы применим фильтр по полю Size и построим регрессионную модель Reviews от Installs**

In [41]:
print(q.Reviews, q.Installs, q.Size)

(0          159
1          967
2        87510
3       215644
4          967
5          167
6          178
7        36815
8        13791
9          121
10       13880
11        8788
12       44829
13        4326
14        1518
15          55
16        3632
17          27
18      194216
19      224399
20         450
21         654
22        7699
23          61
24         118
25         192
26       20260
27         203
28         136
29         223
         ...  
5787      4546
5788         1
5789        21
5790       924
5791       569
5792         1
5793        18
5794        29
5795         8
5796         1
5797        52
5798         0
5799        17
5800         0
5801        56
5802     15806
5803         2
5804         0
5805       440
5806        69
5807         1
5808        94
5809         2
5810         1
5811        53
5812         2
5813       291
5814        71
5815        53
5816         1
Name: Reviews, Length: 5817, dtype: int64, 0           10,000+
1          500,000+
2

# **Анализ данных показывает, что Installs является текстовым полем, которое содержит разделители "," и индикатор "+" в конце строки. Необходимо преобразовать значения из строковых в числовые**

In [42]:
#Убрали последний символ в строке
q.Installs= q.Installs.map(lambda x: str(x)[:-1])
#Убрали ',' в строке
q.Installs = [x.replace(',', '') for x in q.Installs]
q.Installs

0          10000
1         500000
2        5000000
3       50000000
4         100000
5          50000
6          50000
7        1000000
8        1000000
9          10000
10       1000000
11       1000000
12      10000000
13        100000
14        100000
15          5000
16        500000
17         10000
18       5000000
19      10000000
20        100000
21        100000
22        500000
23        100000
24         50000
25         10000
26        500000
27        100000
28         10000
29        100000
          ...   
5787      100000
5788         100
5789        5000
5790      100000
5791       50000
5792         100
5793        1000
5794        1000
5795          50
5796         500
5797        5000
5798          10
5799        5000
5800           5
5801       10000
5802     1000000
5803          10
5804          10
5805       50000
5806       50000
5807          50
5808       10000
5809         100
5810         100
5811        1000
5812           1
5813      100000
5814       100

# **Уберем пустые значение в полях Installs и Reviews**

In [43]:
q[q.Installs=='']
q=q.dropna()#Удаляем все записи с пустыми значениями
q

index  ...         Android Ver
0         0  ...        4.0.3 and up
1         1  ...        4.0.3 and up
2         2  ...        4.0.3 and up
3         3  ...          4.2 and up
4         4  ...          4.4 and up
5         5  ...          2.3 and up
6         6  ...        4.0.3 and up
7         7  ...          4.2 and up
8         8  ...          3.0 and up
9         9  ...        4.0.3 and up
10       10  ...          4.1 and up
11       11  ...          4.0 and up
12       12  ...          4.1 and up
13       13  ...          4.4 and up
14       14  ...          2.3 and up
16       16  ...          4.1 and up
17       17  ...          2.3 and up
18       18  ...        4.0.3 and up
19       19  ...          4.1 and up
20       20  ...          4.1 and up
21       21  ...        4.0.3 and up
22       22  ...        4.0.3 and up
24       24  ...          4.1 and up
25       25  ...          3.0 and up
26       26  ...        4.0.3 and up
27       27  ...        4.0.3 and up
28       28  ...          2.3 and up
29       29  ...          2.3 and up
30       30  ...        4.0.3 and up
31       31  ...          2.3 and up
...     ...  ...                 ...
5775   5775  ...          3.0 and up
5776   5776  ...          4.1 and up
5777   5777  ...          4.3 and up
5778   5778  ...  Varies with device
5779   5779  ...  Varies with device
5780   5780  ...          5.0 and up
5781   5781  ...        2.3.3 and up
5782   5782  ...          4.1 and up
5783   5783  ...        2.3.3 and up
5784   5784  ...          2.2 and up
5785   5785  ...          4.4 and up
5786   5786  ...          4.1 and up
5787   5787  ...          4.0 and up
5789   5789  ...          4.1 and up
5790   5790  ...          3.0 and up
5791   5791  ...          3.0 and up
5794   5794  ...          4.2 and up
5795   5795  ...          4.1 and up
5797   5797  ...          4.1 and up
5799   5799  ...          4.4 and up
5801   5801  ...          3.1 and up
5802   5802  ...          4.1 and up
5803   5803  ...          4.1 and up
5805   5805  ...          2.1 and up
5806   5806  ...        4.0.3 and up
5808   5808  ...          4.3 and up
5811   5811  ...        2.3.3 and up
5813   5813  ...          4.1 and up
5814   5814  ...          2.3 and up
5815   5815  ...        4.0.3 and up

[5456 rows x 14 columns]

# **Меняем тип поля с строки в float**

In [44]:
q.Installs=q.Installs.astype(float)
q.Installs

0          10000.0
1         500000.0
2        5000000.0
3       50000000.0
4         100000.0
5          50000.0
6          50000.0
7        1000000.0
8        1000000.0
9          10000.0
10       1000000.0
11       1000000.0
12      10000000.0
13        100000.0
14        100000.0
16        500000.0
17         10000.0
18       5000000.0
19      10000000.0
20        100000.0
21        100000.0
22        500000.0
24         50000.0
25         10000.0
26        500000.0
27        100000.0
28         10000.0
29        100000.0
30        100000.0
31         50000.0
           ...    
5775        1000.0
5776         100.0
5777        1000.0
5778     1000000.0
5779      100000.0
5780       10000.0
5781      100000.0
5782      500000.0
5783       10000.0
5784     1000000.0
5785       10000.0
5786       10000.0
5787      100000.0
5789        5000.0
5790      100000.0
5791       50000.0
5794        1000.0
5795          50.0
5797        5000.0
5799        5000.0
5801       10000.0
5802     100

# **Также в результате анализа было обнаружено, что поле Size - текстовое и содержит в конце строки буквы "k", "M"(килобайты, мегабайты). Используем функцию для конвертации данных этого поля в один формат (килобайты).**

In [45]:
q.Size = q.Size.map(lambda x: float(str(x)[:-1])*1000 if x[-1] == 'M'  else (float(str(x)[:-1]) if x[-1] == 'k' else x))
q.Size

0                    19000
1                    14000
2                     8700
3                    25000
4                     2800
5                     5600
6                    19000
7                    29000
8                    33000
9                     3100
10                   28000
11                   12000
12                   20000
13                   21000
14                   37000
16                    5500
17                   17000
18                   39000
19                   31000
20                   14000
21                   12000
22                    4200
24                   23000
25                    6000
26                   25000
27                    6100
28                    4600
29                    4200
30                    9200
31                    5200
               ...        
5775                  1600
5776                 24000
5777                  5900
5778    Varies with device
5779    Varies with device
5780                 44000
5

# **Также было обнаружено, что поле Size имеет нечисловые данные. Применим фильтр для удаления таких записей**

In [0]:

q = q[q.Size.map(lambda x: type(x) == float)]

# **Применим фильтр для отбора записей, у которых значение поле Size больше 20М (20000 кб)**

In [59]:
q = q[q['Size']>=20000]
q

index  ...   Android Ver
3         3  ...    4.2 and up
7         7  ...    4.2 and up
8         8  ...    3.0 and up
10       10  ...    4.1 and up
12       12  ...    4.1 and up
13       13  ...    4.4 and up
14       14  ...    2.3 and up
18       18  ...  4.0.3 and up
19       19  ...    4.1 and up
24       24  ...    4.1 and up
26       26  ...  4.0.3 and up
36       36  ...    4.4 and up
46       45  ...    4.1 and up
47       46  ...    4.1 and up
50       49  ...  4.0.3 and up
51       50  ...    4.1 and up
52       51  ...    4.1 and up
54       53  ...    4.1 and up
55       54  ...    4.1 and up
56       55  ...    4.0 and up
58       57  ...    4.1 and up
65       64  ...    4.2 and up
71       70  ...  4.0.3 and up
80       79  ...    4.1 and up
81       80  ...    6.0 and up
82       81  ...    4.2 and up
87       86  ...    2.3 and up
92       91  ...  4.0.3 and up
102     101  ...  4.0.3 and up
104     103  ...    4.1 and up
...     ...  ...           ...
5702   5702  ...    4.1 and up
5703   5703  ...    4.1 and up
5714   5714  ...    4.1 and up
5717   5717  ...    4.0 and up
5719   5719  ...    4.1 and up
5720   5720  ...    4.1 and up
5725   5725  ...    4.2 and up
5726   5726  ...    4.1 and up
5728   5728  ...    3.0 and up
5731   5731  ...    4.1 and up
5733   5733  ...    4.0 and up
5743   5743  ...    5.0 and up
5746   5746  ...    2.3 and up
5761   5761  ...    4.0 and up
5764   5764  ...  4.0.3 and up
5766   5766  ...    4.1 and up
5774   5774  ...    4.3 and up
5776   5776  ...    4.1 and up
5780   5780  ...    5.0 and up
5781   5781  ...  2.3.3 and up
5782   5782  ...    4.1 and up
5785   5785  ...    4.4 and up
5790   5790  ...    3.0 and up
5794   5794  ...    4.2 and up
5795   5795  ...    4.1 and up
5797   5797  ...    4.1 and up
5802   5802  ...    4.1 and up
5808   5808  ...    4.3 and up
5813   5813  ...    4.1 and up
5814   5814  ...    2.3 and up

[1951 rows x 14 columns]

# **Сделаем преобразование данных (нормализацию)**

In [0]:
q_s=q #используем датафрейм q_s с нормализованными данными
q_s.Installs=q_s.Installs.transform(lambda x: (x / x.max()))

In [61]:
q_s.Installs.max()

1.0

In [62]:
q_s.Installs

3       5.000000e-02
7       1.000000e-03
8       1.000000e-03
10      1.000000e-03
12      1.000000e-02
13      1.000000e-04
14      1.000000e-04
18      5.000000e-03
19      1.000000e-02
24      5.000000e-05
26      5.000000e-04
36      5.000000e-04
46      1.000000e-02
47      1.000000e-04
50      1.000000e-04
51      1.000000e-03
52      1.000000e-04
54      1.000000e-04
55      1.000000e-04
56      1.000000e-04
58      1.000000e-04
65      1.000000e-04
71      5.000000e-03
80      1.000000e-03
81      5.000000e-05
82      1.000000e-03
87      1.000000e-04
92      1.000000e-03
102     5.000000e-04
104     5.000000e-04
            ...     
5702    1.000000e-02
5703    1.000000e-06
5714    1.000000e-02
5717    1.000000e-05
5719    5.000000e-02
5720    1.000000e-02
5725    1.000000e-06
5726    1.000000e-05
5728    1.000000e-02
5731    1.000000e-02
5733    1.000000e-07
5743    1.000000e-04
5746    1.000000e-03
5761    5.000000e-03
5764    1.000000e-04
5766    1.000000e-02
5774    1.000

In [63]:
q_s.Reviews=q_s.Reviews.transform(lambda x: (x / x.max()))
q_s.Reviews

3       4.803416e-03
7       8.200448e-04
8       3.071910e-04
10      3.091735e-04
12      9.985546e-04
13      9.636056e-05
14      3.381307e-05
18      4.326112e-03
19      4.998431e-03
24      2.628420e-06
26      4.512864e-04
36      1.814278e-04
46      3.887634e-03
47      2.383398e-05
50      8.174832e-06
51      3.559505e-05
52      6.326028e-06
54      2.873442e-06
55      1.207291e-05
56      2.334171e-04
58      2.873442e-06
65      1.356532e-04
71      2.605834e-03
80      4.391155e-03
81      3.163014e-06
82      3.378634e-04
87      7.751612e-06
92      1.070524e-04
102     3.873579e-05
104     4.956131e-05
            ...     
5702    2.840861e-02
5703    9.578141e-07
5714    2.607573e-02
5717    5.769159e-06
5719    5.788269e-02
5720    1.280172e-02
5725    5.345939e-07
5726    2.027002e-06
5728    2.128606e-02
5731    1.444250e-02
5733    8.909899e-08
5743    1.719165e-04
5746    6.473487e-04
5761    4.671683e-04
5764    1.095918e-05
5766    3.573195e-02
5774    2.574

#**Создаем функцию для построения регрессионной модели на основе sklearn.linear_model**

In [0]:
import pickle
def model_(data_model):
   
    Y = data_model.iloc[:,[4]] # Reviews - 4 поле
    X=data_model.iloc[:,[6]]#Installs -6 поле
    
    filename = 'Model.sav'
    
    model = LinearRegression().fit(X,Y) #Обучение модели
    r_sq = model.score(X,Y) #Оценка качества модели
    print(r_sq)
    print(model.predict(X[0:1]))#Прогнозирование для 1-ого наблюдения
    pickle.dump(model, open(filename, 'wb')) #Сохранение модели на диске

#**Моделируем**

In [65]:
model_(q)

0.48292590819291736
[[0.03882758]]


# **Результаты моделирования (R^2 = 0.4829) показывают наличие некоторой зависимости между Reviews и Installs, но она не очень большая**

In [66]:
model_(q_s)#Это справедливо и для нормализованных данных

0.48292590819291736
[[0.03882758]]


# **Пример очистки на основе собственных функций**

In [31]:
df=pd.read_sql_query('''SELECT * FROM entrexams;''', cnx)#таблица entrexams содержит данные для примера
df

exam1  exam2  exam3  exam4 studentcode
0     98     97     96     99           1
1     90     90     90     99           2
2     98     88     74     80           3
3     70     60     74     75           5
4     80     90     74     84           6
5     87     76     77     80           4
6     92     90     80     78           7
7     71     68     60     60           8
8     73     74     85     88           9
9     80     83     85     88          10

# **Функция возвращает результат в зависимости от стартовых значений поля exam1**

In [0]:
def calc_(exam):
    
    if exam>=90:
          result = 'A'

    elif exam>=80: 
          result = 'B'
          
    elif exam>=74: 
          result = 'C'
    else: 
          result = 'FF'
    return result 

# **Использования apply для вызова функции для каждого значения поля 'exam1'**

In [32]:
df['exam1'] = df.apply(lambda row: calc_(row['exam1']), axis=1)
df.exam1

0     A
1     A
2     A
3    FF
4     B
5     B
6     A
7    FF
8    FF
9     B
Name: exam1, dtype: object

In [33]:
df.loc[df['exam1'].isin(['A'])]# Поиск, если поле содержит символ A

exam1  exam2  exam3  exam4 studentcode
0     A     97     96     99           1
1     A     90     90     99           2
2     A     88     74     80           3
6     A     90     80     78           7